filebeat/reader/line/line.go

```diff
@@ -18,177 +18,258 @@
 package line
 
 import (
+	"fmt"
 	"io"
 
 	"golang.org/x/text/encoding"
 	"golang.org/x/text/transform"
 
+	"github.com/elastic/beats/libbeat/common"
 	"github.com/elastic/beats/libbeat/common/streambuf"
-	"github.com/elastic/beats/libbeat/logp"
 )
 
 // lineReader reads lines from underlying reader, decoding the input stream
 // using the configured codec. The reader keeps track of bytes consumed
 // from raw input stream for every decoded line.
 type Reader struct {
-	reader     io.Reader
-	codec      encoding.Encoding
-	bufferSize int
-	nl         []byte
-	inBuffer   *streambuf.Buffer
-	outBuffer  *streambuf.Buffer
-	inOffset   int // input buffer read offset
-	byteCount  int // number of bytes decoded from input buffer into output buffer
-	decoder    transform.Transformer
+	lineScanner *lineScanner
 }
 
 // New creates a new reader object
-func New(input io.Reader, codec encoding.Encoding, bufferSize int) (*Reader, error) {
-	encoder := codec.NewEncoder()
-
-	// Create newline char based on encoding
-	nl, _, err := transform.Bytes(encoder, []byte{'\n'})
-	if err != nil {
-		return nil, err
-	}
+func New(input io.Reader, codec encoding.Encoding, separator []byte, bufferSize int) *Reader {
+	decReader := newDecoderReader(input, codec, bufferSize)
+	lineScanner := newLineScanner(decReader, separator, bufferSize)
 
 	return &Reader{
-		reader:     input,
-		codec:      codec,
-		bufferSize: bufferSize,
-		nl:         nl,
-		decoder:    codec.NewDecoder(),
-		inBuffer:   streambuf.New(nil),
-		outBuffer:  streambuf.New(nil),
-	}, nil
+		lineScanner: lineScanner,
+	}
 }
 
 // Next reads the next line until the new line character
 func (r *Reader) Next() ([]byte, int, error) {
-	// This loop is need in case advance detects an line ending which turns out
-	// not to be one when decoded. If that is the case, reading continues.
-	for {
-		// read next 'potential' line from input buffer/reader
-		err := r.advance()
-		if err != nil {
-			return nil, 0, err
-		}
+	return r.lineScanner.scan()
+}
 
-		// Check last decoded byte really being '\n' also unencoded
-		// if not, continue reading
-		buf := r.outBuffer.Bytes()
+type decoderReader struct {
+	in         io.Reader
+	decoder    transform.Transformer
+	buf        *streambuf.Buffer
+	encodedBuf *streambuf.Buffer
+	bufferSize int
+	symlen     []int
```

Comment:
let's safe some space and use `[]uint8`. On 64bit architectures an int takes 8bytes.